### PR TITLE
Feature/pr83 integration

### PR DESCRIPTION
**Pull request by bvarner**
Creating a pull request with my current work toward integrating / resolving issues encountered while pulling in items from #83 .

This is a WIP, but I wanted to start the conversation and invite some contributions over the weekend.

Basically this pulls in everything from the prior PR, but takes it a step further to refactor some structures, a few logic bits that were highly confusing at first glance to me, and restore some functionality that was broken.

I'm dealing with code bases that *don't* have `<scm>` blocks. Yep, I don't like that either, but it's the way it is.

* ScmUtils refactoring....
    * was really wonky as a stateless utils class. I should consider this to need renamed as a helper now, but the API and the amount of arguments is much cleaner.
    * Rolled up the git branch detection to be a bit less obtuse. I am leery of the amount of code still in that class.
    * Restored the ability to resolve the branch name or the expresion to use for resolving a branch name (see above about building projects without <scm> blocks). Previous commits completely obliterated the property resolver on this, and you cannot specify a resolvable property as an `@Parameter` and it work the way you'd want.
* Refactored the entry points to the Mojos to use GitBranchInfo as the argument, instead of constatly decomposing this and passing it about.

Things still which should be looked at:

* Perhaps rename GitBranchInfo to GitBranchContext, and further clean up the resolution a bit.

### CODE REVIEW

**Comment by robth** [comment]
Hi @bvarner, sorry for the late response. Let me know if I can help you with anything!

**Comment by roband7** [comment]
It would be great if this could be pushed into a release soon, as unfortunately the project I'm currently working on has some resistance to non-released patches.

**Comment by bvarner** [comment]
Totally understand that. Should we merge as-is? Have you been able to test this at all?

**Comment by roband7** [comment]
I'm afraid I've only done code review and superficial testing. I don't mind merging as-is.

I wish I could have done more but I've been waiting for the 2018.2 release of IntelliJ IDEA that has some other fixes I also need. It's the first release that supports https://maven.apache.org/maven-ci-friendly.html

**Comment by robth** [comment]
Thanks for all the hard work, @bvarner!

I agree a release would be great. It would be even better if my small change from PR #85 can be integrated in the release. It ensures that, for a release type branch, the project itself isn't a SNAPSHOT, but it parents aren't too.

File: src/main/java/com/e_gineering/maven/gitflowhelper/EnforceVersionsMojo.java [hunk 1]
```diff
@@ -1,49 +1,145 @@
...
 /**
  * If there is an ${env.GIT_BRANCH} property, assert that the current ${project.version} is semantically correct for the
- * git branch.
+ * git branch. Also, make sure there are no SNAPSHOT (plugin) dependencies.
  */
-@Mojo(name = "enforce-versions", defaultPhase = LifecyclePhase.VALIDATE)
+@Mojo(requiresDependencyCollection = ResolutionScope.TEST, name = "enforce-versions", defaultPhase = LifecyclePhase.VALIDATE)
 public class EnforceVersionsMojo extends AbstractGitflowBranchMojo {
 
+    @Parameter(defaultValue = "true", property = "enforceNonSnapshots", required = true)
```
Comment (jacobharris919) [inline src/main/java/com/e_gineering/maven/gitflowhelper/EnforceVersionsMojo.java]: Documentation needed for this option.  I had to look at the code to find the name of the property to set when I wanted to turn this off.

**Comment by bvarner** [comment]
My sincere apologies for the delays, and much thanks to @jacobharris919 for helping test things out yesterday. This long overdue merge is in, and I'm cutting release 1.8.0 now.

**Pull request merged**